### DISCUSSION TITLE
New task. Textual Anomaly Detector.

### DISCUSSION THREAD

**Discussion by kennynakamura** [sepinf-inc/IPED#508] [Ideas]
Would a textual anomaly detector be interesting? Find phrases outside the context of a conversation that could represent secret codes.

https://github.com/kennynakamura/Detector-de-Anomalias

**Comment by lfcnassif** [comment]
For sure. What is the license of the suggested library? Did you evaluate the number of false positives and negatives of the proposed approach on some dataset?

**Reply by kennynakamura** [reply]
The two libraries I used are open source (spacy and bert). 

I created an example of an email with hidden codes (In the repostory has the test and result of processing). 
[EmailTeste.txt](https://github.com/sepinf-inc/IPED/files/6417979/EmailTeste.txt)
[(BERT)EmailTeste.txt - Resultado -C1-F0.5.txt](https://github.com/sepinf-inc/IPED/files/6417980/BERT.EmailTeste.txt.-.Resultado.-C1-F0.5.txt)

The algorithm I proposed separates approximately 5% of the text with the most out of context phrases. It is difficult to separate exactly the phrases that are codes.
I do not know for sure how to evaluate, but there are always false positives because I don't put 5% of the text with codes and and the results vary according to the text.

thanks for answering! Sorry for my English, he is only good for read Stackoverflow kkk

**Reply by kennynakamura** [reply]
I developed with two methods, one with each library (spacy and bert). The logic of the two methods is the same. They compare the probability that two sentences are logically related. But as they are different libraries, the results are different. The spacy method is faster, but I believe it is less accurate. The bert method has had better results so far, but  need more robust libraries and is considerably slower. 
They are widely used in natural language processes and can be accessed offline.

**Reply by lfcnassif** [reply]
This is a really interesting work and with a big potential for helping investigators and examiners to analyze lots of data.

Just a note, not all open source licenses are compatible with each other, so not all are compatible with IPED license. I took a look at spacy and bert licenses, they are MIT based and Apache v2 repectively, so they are ok to be used. But not sure about the license of your code.

And a question: is it possible to, instead of fixing 5% of any text to be tagged as outlier, to set a dissimilarity threshould? And that should be set previously or could be set at analysis time?

**Reply by kennynakamura** [reply]
Thank you! I'm glad you liked it.

Didn’t know about these license specifications, I updated mine. I put MIT license. Thanks

When starting the program, it is set a dissimilarity threshould. The value is between 0.1 to 0.99 (almost as a percentage), more lower the value, more chances of being a outliner. Therefore, any comparisons that result below the threshould value, have their phrases tagged as outliner. The program makes comparisons between all possible permutations of part (I'll explain later) of analyzed text. 

![fluxogramaalgoritmo](https://user-images.githubusercontent.com/8608969/117025735-8c7bd480-acd1-11eb-9459-a221043ddac3.png)

Many comparisons that are considered normal for us, are evaluate as below the threshould and are tagged as outliner. Because of this many phrases will be tagged outliner, so the number of times a sentence is tagged as outliner a ranking is made. The best placed are selected at the end.

The analyzed text is divided into parts due to the processing time. The more lines the more permutation possibilities exist and the more comparisons are made. And also because the results were good.

Example:
analyzed text with 1000 lines  ----> 50 parts of original text (20 lines each part)  --->  for each part, 1 line is selected off the ranking made.
At the end we will have 50 sentences, 5% of original text.

In texts with other numbers of lines, the values ​​are rounded, so approximately 5% of the text are selected.
It is possible to change the 5%, but I think it is the value that best optimizes the analysis.
Less than that the chances of selecting are less. And more than 5% more phrases will be selected.

if something is not clear, please let me know. Eu posso explicar melhor em português, se for o caso.

**Reply by lfcnassif** [reply]
I think I understood the general approach. Is there some kind of sentence boundary detection or text is always broken by lines? And could a second ranked line in one block have a better score than the top ranked line of another block? Maybe allowing the user to set the minimal ranking score could improve a bit?

Anyway, things can always be enhanced later and I think a first version could be very helpful.

**Reply by lfcnassif** [reply]
PS: Thanks for clarifying the license :)

**Reply by kennynakamura** [reply]
Most of the code is data cleanning. I did it in a rustic way for not knowing so well regex, but the main process is to separate the sentences and remove punctuation. That's because long sentences don't work well. Separating them is better.

**Example:**
Original:
Na primeira semana, meus primos foram lá em casa, mas como estava chovendo, nós não pudemos sair para brincar!
Modified:
na primeira semana
meus primos foram lá em casa
mas como estava chovendo
nós não pudemos sair para brincar

There are also processes to filter only the texts of whatsapp messages. In this case the messages have many Portuguese errors, abbreviations, laughter, etc. But I had tests that within the 5% was the message with code that I added to the conversation.

**Example**
Original:
10/12/20 19:42 - Kenny: Opa kkk, de Sábado pode ser?
Modified:
Opa
de sábado pode ser

Yes, it's possible. Good idea!! I hadn't thought that. I will test this, but I already have a good feeling. 

In code there is also parallel processing by CPU. 
With parallel processing by CPU I got reasonable time process, but I don't know if with real cases this process time is acceptable.
BERT model = 1000 lines in less than 20 minutes
Spacy model = 1000 lines in less than 5 minutes
16gb RAM, i7, 4 cores in parallel
I don't know if a GPU is necessary, as these are texts, but I also have no experience in GPU inference

And it also has a loading screen in a other thread (Carregando ... )

If you think that any of these processes is not necessary or may make it difficult to function in IPED, I can change the code.
I am studying the code of FaceDtection to understand how to apply the python code in the program in java too, if there is something to change, please let me know.

**Reply by kennynakamura** [reply]
PS: Each block is separate after data cleannig
Example:
Original Text = 870 lines
After data cleannig (separate sentences, etc) = 1000 lines 
so the whole process

**Reply by lfcnassif** [reply]
You can also take a look at Yahoo NSFW based nudity detection: https://github.com/sepinf-inc/IPED/blob/master/iped-app/resources/config/conf/scripts/NSFWNudityDetectTask.py

It does not open external processes (simpler) but does process items in batches (more complex).

**Comment by kennynakamura** [comment]
I did some tests with the possibility of the user choosing the minimal ranking score, but the range is minimal to choose without many lines being selected. I did the tests with whatsapp group messages from university students, things tend to be a bit "chaotic" too, many messages are really strange. In email with a formal language the results were neither better nor worse.

I was looking at the python scripts, there are "FaceRecognitionFaceProcess.py" and "FaceRecognitionFaceTask.py", the script "Process", I understood. But o "Task"  I didn't understand it very well, my knowledge in Python + Java is almost none. 
To make a pull request, you need both? 

The spacy model can be accessed offline, but needs to be downloaded previously, or it would be better to leave access to it online?

I am studying about the batches!

**Reply by lfcnassif** [reply]
Sorry about late reply, lots of things happening behind the scenes...

Just a ...Task.py script is needed. Take a look at the basic example, it explains some things about the basic Task structure and how to enable it: https://github.com/sepinf-inc/IPED/blob/master/iped-app/resources/config/conf/scripts/PythonScriptTask.py

Offline processing is better for sure.

I would not recommend trying to do batch processing of items for now, things can become complicated and buggy. I just pointed the NSFWNudityDetectTask.py because it is a single script, I should have pointed the basic PythonScriptTask.py before.

**Reply by kennynakamura** [reply]
No problem, i thank you for your attention. Good luck out there!

I'm reading "PythonScriptTask.py", realized that it is in Object-oriented programming wich I not much familiar. I realized that it's about the iped software part, right?
In "FaceRecognitionFaceTask.py" and inside off Class is a **"def process(self, item):"** which is common in "PythonScriptTask.py" too, and the python script process is called over there. But i'm lost on how to write all the code. Can you help me with that?
Sorry for not knowing a lot about how to adapt the code.

If I learn well how to do all the parts, I can create other features that use the same spacy lib. I was thinking about a threat detector in conversations.

**Reply by lfcnassif** [reply]
The main method is **def process(...)** as you noticed, in that method you can access the decoded text of files calling item.getParsedTextCache() like shown at the end of the basic example, so you can use your code to process item text and store the results like shown in the last line of the example.

Early versions of python tasks were not object oriented, but that was changed to easy the implementation of thread safe code, your code will be called by several threads at the same time by iped engine.

**Reply by lfcnassif** [reply]
You should write all your code in one task script file. FaceRecognitonTask was broken in two scripts to be executed in external processes because of an infamous global lock bottleneck of CPython, see the details on #361. Some smarter libs take care of releasing and acquiring the GIL internally when crossing python<->C boundaries, like tensorflow, so external processes would not be needed, ideally they shouldn't...

**Comment by kennynakamura** [comment]
https://github.com/kennynakamura/Detector-de-Anomalias/blob/main/AnomalyDetectorTask.py
https://github.com/kennynakamura/Detector-de-Anomalias/blob/main/AnomalyDetectorConfig.txt
https://github.com/kennynakamura/Detector-de-Anomalias/blob/main/EmailTeste_Resultado_An%C3%A1lise.txt

Hello! 
Just an update to the code i'm writing.
In the **def process(..)** I put almost all the code, in **def init(..)**  i put the model loading and parameters (filtro) from Config file.
The model was download from here https://spacy.io/models/pt#pt_core_news_lg 
When testing in IPED, putting it in the **scripts folder** and activating in **XML** , it works, but I'm in doubt where to save the analysis made. 
I commented on the code, but if something is unclear, please ask me!

**Reply by lfcnassif** [reply]
Thanks @kennynakamura! I will take a look when some time is available.

**Reply by lfcnassif** [reply]
Hi @kennynakamura you can store the results in a new property into the item:
item.setExtraAttribute('possibleTextAnomaly', string)

second param can also be a string list with multiple text chunks.

**Reply by kennynakamura** [reply]
![Captura de Tela (1)](https://user-images.githubusercontent.com/8608969/119374186-aafd3c00-bc8f-11eb-970f-71a43ba6e062.png)

https://github.com/kennynakamura/Detector-de-Anomalias/blob/main/AnomalyDetectorTask.py

I updated the code as suggested, I think this is it, right?

Results may vary depending on the filter value (filter_det in config.txt) and the analyzed text. For the LG spacy model, a value between 0.2 ~ 0.5 is recommended. 
it works with whatsapp conversations, but needs to do a datacleanning of only the messages, without date, sender, etc.

**Reply by lfcnassif** [reply]
Cool. You can also check the results in the Metadata Viewer, second subtab, if I'm not wrong.

I have 2 concerns:
1. If I understood correctly, all texts will have text chunks tagged as anomaly, that should be true Just for a few.
2. Processing time in common cases with hundreds of thousands of items. The algorithm seems quadratic, so you break texts in blocks to workaround it. Not sure about the final cost it takes relative to current processing time.

**Reply by kennynakamura** [reply]
I think that few files would really be interesting for analysis, so I was thinking of using it as a tool on selected specific files after the IPED-SearchApp created, like some tools in the options (ex. find similar images), then the processing time would not be that long. Run on all files in image.dd file, can greatly increase processing time. 

I'm sorry if I'm saying something silly, I didn't use it professionally, having no experience in real cases.

**Reply by lfcnassif** [reply]
That's a good idea. But "Find Similar Images" processes all images before to enable image similarity search in UI.

Maybe #24 (scheduled to be implemented soon) would help running this on user selected files.

**Reply by kennynakamura** [reply]
I saw that in **CategoriesByTypesConfig.txt** item mimetypes are classified. If just "chat" and "email" items were process by my script? 

The #24 it looks very promising!

**Reply by lfcnassif** [reply]
Yes, this is another approach, "Find similar images" process just images before hand.
Try something like this at the beginning of the process(self, item) method:

```
categories = item.getCategorySet().toString()
if not ("Chats" in categories or "Emails" in categories):
    return
```

**Reply by kennynakamura** [reply]
I'm testing, but I'm not getting text, is **item.getParsedTextCache()** works normally for EML file?
If i do
`texto = str(item.getParsedTextCache())`
`print(texto)`

Just METADADOS are print in log and in SearchApp

**Reply by lfcnassif** [reply]
Yes, it should work. Your script should be installed after ParsingTask in TaskInstaller.xml file to be able to get decoded text. Not sure why it is not being printed in IPEDSearchApp.exe, could you elaborate?

**Reply by lfcnassif** [reply]
Can be related to a recent upgrade of Tika lib. Will test some emls here. If you could share your eml file, that Will help.

**Reply by kennynakamura** [reply]
My eml file was wrong, I corrected and works! 

![Captura de Tela (9)](https://user-images.githubusercontent.com/8608969/119538974-0861bd80-bd62-11eb-9d92-06c4f88ed7b7.png)

I update my code:
https://github.com/kennynakamura/Detector-de-Anomalias/blob/main/AnomalyDetectorTask.py

**Reply by lfcnassif** [reply]
Great! But I think running over all emls (could be hundreds of thousands) and chats will result in lots of false positives, as potential anomalies will be extracted from all of them, even those without anomalies (most of them). So, without #24, I think the algorithm should tag just some phrases of some (not all) emails...

**Reply by kennynakamura** [reply]
Yes I agree. It works well on large suspicious emails, for all the emails available, I don't think it would really worth it.

**Reply by lfcnassif** [reply]
Or another approach is running other algorithm before to tag suspicious emails, then just those would be submitted to this algorithm...

**Reply by kennynakamura** [reply]
I saw that there is a function of clustering files (_**Powerful file grouping (clustering) based on ANY metadata**_), if the email cluster with the least amount is analyzed? but it would probably contain a lot of emails yet, just something that I tried to think.

**Reply by lfcnassif** [reply]
It is a bit overstated. That phrase refers exactly to Metadata filter panel located at the bottom-left of the GUI, it groups results based on metadata and allow filtering by them.

**Reply by kennynakamura** [reply]
I wait for the #24 or, if you think my algorithm is fine, i can make a pull-request?

**Reply by lfcnassif** [reply]
I think this will need to wait #24 because IPED main purpose is to process large amounts of files and, as I understood, currently this algorithm assumes the texts always have anomalies, which may be true just for a few cases. If someone else has a better idea...
